src/goto-instrument/contracts/contracts.cpp

```diff
@@ -133,34 +137,66 @@ void code_contractst::check_apply_loop_contracts(
     parameter2history,
     loop_head->source_location(),
     mode,
-    havoc_code,
+    generated_code,
     ID_loop_entry);
 
   // Create 'loop_entry' history variables
-  insert_before_swap_and_advance(goto_function.body, loop_head, havoc_code);
+  insert_before_swap_and_advance(goto_function.body, loop_head, generated_code);
 
   // Generate: assert(invariant) just before the loop
   // We use a block scope to create a temporary assertion,
   // and immediately convert it to goto instructions.
   {
     code_assertt assertion{invariant_expr()};
     assertion.add_source_location() = loop_head->source_location();
-    converter.goto_convert(assertion, havoc_code, mode);
-    havoc_code.instructions.back().source_location_nonconst().set_comment(
+    converter.goto_convert(assertion, generated_code, mode);
+    generated_code.instructions.back().source_location_nonconst().set_comment(
       "Check loop invariant before entry");
   }
 
   // havoc the variables that may be modified
-  havoc_if_validt havoc_gen(modifies, ns);
-  havoc_gen.append_full_havoc_code(loop_head->source_location(), havoc_code);
+  modifiest modifies;
+  if(assigns.is_nil())
+  {
+    try
+    {
+      get_modifies(local_may_alias, loop, modifies);
+    }
+    catch(const analysis_exceptiont &exc)
+    {
+      log.error() << "Failed to infer variables being modified by the loop at "
+                  << loop_head->source_location()
+                  << ".\nPlease specify an assigns clause.\nReason:"
+                  << messaget::eom;
+      throw exc;
+    }
+  }
+  else
+  {
+    for(const auto &target : assigns.operands())
+      modifies.insert(target);
+
+    // create snapshots of the CARs -- must be done before havocing
+    for(const auto &car : loop_assigns.get_write_set())
+    {
+      auto snapshot_instructions = car.generate_snapshot_instructions();
+      insert_before_swap_and_advance(
+        goto_function.body, loop_head, snapshot_instructions);
+    };
+  }
+
+  havoc_assigns_targetst havoc_gen(modifies, ns);
+  havoc_gen.append_full_havoc_code(
+    loop_head->source_location(), generated_code);
+  insert_before_swap_and_advance(goto_function.body, loop_head, generated_code);
 
   // Generate: assume(invariant) just after havocing
   // We use a block scope to create a temporary assumption,
   // and immediately convert it to goto instructions.
   {
     code_assumet assumption{invariant_expr()};
     assumption.add_source_location() = loop_head->source_location();
-    converter.goto_convert(assumption, havoc_code, mode);
+    converter.goto_convert(assumption, generated_code, mode);
   }
```

Comment:
As above: likely doesn't actually require the use of `goto_convert`?

Reply:
As discussed offline this morning, `goto_convert` is necessary here because the `assume(invariant)` statement could expand to multiple GOTO instructions.